### ISSUE TITLE
Node Adapter Not Supporting URLs With #anchors

### ISSUE THREAD

**Issue by porl** [withastro/astro#9718]
### Astro Info

```block
Local (dev machine):
Astro                    v4.0.6
Node                     v18.18.2
System                   Linux (x64)
Package Manager          pnpm
Output                   hybrid
Adapter                  @astrojs/node
Integrations             @astrojs/tailwind

Server: 
Astro                    v4.0.6
Node                     v18.19.0
System                   Linux (x64)
Package Manager          pnpm
Output                   hybrid
Adapter                  @astrojs/node
Integrations             @astrojs/tailwind
```

### Describe the Bug

Whilst developing locally, the site behaves as expected for links to specific anchors on another page. For example, having a link to `/classes#kids` works as you would expect - scrolls to the element with id `kids` on page `/classes`.

Running in hybrid mode on my production server with Node as the SSR adapter however, the links are still formatted the same `/classes#kids` etc., however as soon as they are clicked, the browser loads the url `/classes/`. Note the trailing `/` and removal of `#kids`.

### What's the expected result?

Hybrid sites running on Node adapter should resolve links to specific ids on a page the same way it does for the local astro dev server.

### Link to Minimal Reproducible Example

https://astrolinks.euclid.net.au

### Participation

- [X] I am willing to submit a pull request for this issue.

**Comment by porl** [comment]
Just noticed the Minimal Reproducible Example page is down. Heading home now but will quickly fix that when I get there.

**Comment by matthewp** [comment]
Can you provide a reproduction through https://astro.new/latest ? A site running in production doesn't help a lot because we can't see the code.

Hashes in URLs are not sent to servers, it's purely a client-side thing. So It's likely a case where the links do not contain the hashes. I'd like to see a reproduction to confirm.

**Comment by porl** [comment]
I am trying to reproduce this problem and am struggling to do so. I created a new project and set it up the same way I did the original site. This time both the local dev and node servers behave as expected.

I noticed I was running an earlier (4.0.x) branch of Astro on the main site so thought maybe there was something there. Updating to latest and now *both* the local dev and node versions have the problem.

Apart from adding Tailwind I'm pretty much using a minimal setup so I'm not sure what the problem is. I know you are after a minimal build but while I try to figure out what that will need I have uploaded my main site code to github here:

https://github.com/porl/judocentralcoast.com.au

The problem is mostly when going from `/timetable` to one of the class links - they should take you to `/classes#[classname]` (scrolling the page down to the relevant section) as you can see in the link hrefs. However the page seems to "redirect" to `/classes/` (with trailing slash) and drop the `#[classname]` part. The only "SSR" page is the /contact one so I would have thought the rest are being served mostly statically in hybrid mode.

Sorry I don't yet have a more minimal example. I am trying to rebuild the functionality but apart from adding tailwind I'm not sure what else is "non-vanilla".

**Comment by ematipico** [comment]
I can't even access the website you posted at the beginning. I get a 404

**Comment by ematipico** [comment]
It's possible that the issue lies in the node adapter

**Comment by porl** [comment]
> I can't even access the website you posted at the beginning. I get a 404

Yes, I'm having problems with that one sorry. I tried to upload a minimal test site to show it live. In the meantime I have the code for the "real" site at the github link above. The actual site is at https://judocentralcoast.com.au which shows the problem.

**Comment by ematipico** [comment]
I downloaded your repository an run it locally:
- when running on dev, I see the desired behaviour: no trailing slash applied and scrolling to hash occurs as expected
- the project doesn't build so I couldn't try the preview command

**Comment by porl** [comment]
>     * the project doesn't build, so I couldn't try the preview command

That's strange, I have no problems building it at all (I'm using pnpm if that makes a difference). Is there anything that could cause this?

**Comment by porl** [comment]
> Ok, I was running `pnpm build`, which runs `astro check` and spits a lot of errors. I run `pnpm astro build` now. With the preview server **I can replicate the issue!**
> 
> That's definitely odd

Never looked at `astro check` before, so didn't realise I had so many errors! Thanks for letting me see that.

I'm glad the issue isn't just happening for me. Let me know if there is anything else I can do to test. As I mentioned, building a new project with the minimal setup doesn't seem to replicate it.

**Comment by ematipico** [comment]
@porl 

I did some digging, and the issue comes from the `trailingSlash`: https://docs.astro.build/en/reference/configuration-reference/#trailingslash

Particularly, the node adapter has some issues in supporting `trailingSlash` at the moment, and we have a PR for that.

As a workaround, I suggest you have all your links with the trailing slash

**Comment by porl** [comment]
Thanks for the workaround suggestion - this works fine for now.

I've left the github repo as-is (I use another host for my main repo) for reference for now. In the meantime I have pushed the change to my server and it works fine - provided I leave the node adapter version back at 7.0.4. See this comment https://github.com/withastro/astro/pull/9661#issuecomment-1901553702 I added to the update merge thread for reference. Not sure why 8.0.0 doesn't like my Traefik setup but at least with version 7.0.4 I have a functioning site.
